pymc3/distributions/simulator.py

```diff
@@ -54,86 +68,102 @@ class Simulator(NoDistribution):
-            self.sum_stat = np.sort
+            sum_stat = at.sort
         elif sum_stat == "mean":
-            self.sum_stat = np.mean
+            sum_stat = at.mean
         elif sum_stat == "median":
-            self.sum_stat = np.median
+            sum_stat = at.median
         elif hasattr(sum_stat, "__call__"):
-            self.sum_stat = sum_stat
+            # TODO: Wrap (optionally) non symbolic sum_stat in Aesara OP
+            sum_stat = sum_stat
         else:
             raise ValueError(f"The summary statistics {sum_stat} is not implemented")
 
-        super().__init__(shape=np.prod(observed.shape), dtype=observed.dtype, *args, **kwargs)
-
-    def random(self, point=None, size=None):
-        """
-        Draw random values from Simulator.
-
-        Parameters
-        ----------
-        point: dict, optional
-            Dict of variable values on which random values are to be conditioned (uses default
-            point if not specified).
-        size: int, optional
-            Desired size of random sample (returns one sample if not specified).
-
-        Returns
-        -------
-        array
-        """
-        # size = to_tuple(size)
-        # params = draw_values([*self.params], point=point, size=size)
-        # if len(size) == 0:
-        #     return self.function(*params)
-        # else:
-        #     return np.array([self.function(*params) for _ in range(size[0])])
-
-    def _str_repr(self, name=None, dist=None, formatting="plain"):
-        if dist is None:
-            dist = self
-        name = name
-        function = dist.function.__name__
-        params = ", ".join([var.name for var in dist.params])
-        sum_stat = self.sum_stat.__name__ if hasattr(self.sum_stat, "__call__") else self.sum_stat
-        distance = getattr(self.distance, "__name__", self.distance.__class__.__name__)
-
-        if "latex" in formatting:
-            return f"$\\text{{{name}}} \\sim  \\text{{Simulator}}(\\text{{{function}}}({params}), \\text{{{distance}}}, \\text{{{sum_stat}}})$"
-        else:
-            return f"{name} ~ Simulator({function}({params}), {distance}, {sum_stat})"
+        if params is None:
+            params = []
+
+        # Assume scalar ndims_params
+        if ndims_params is None:
+            ndims_params = [0] * len(params)
+
+        sim_op = type(
+            f"Simulator_{name}",
+            (SimulatorRV,),
+            dict(
+                name="Simulator",
+                ndim_supp=ndim_supp,
+                ndims_params=ndims_params,
+                dtype=dtype,
+                inplace=False,
+                # Specifc to Simulator
+                fn=fn,
+                # distance=distance,
+                # sum_stat=sum_stat,
+                # epsilon=epsilon,
+            ),
+        )()
+
+        # Register custom logp
+        rv_type = type(sim_op)
+
+        @_logp.register(rv_type)
+        def logp(op, sim_rv, rvs_to_values, *sim_params, **kwargs):
```

Comment:
locally defined functions often cause problems with pickling.

Also it looks like this `logp` uses variables from the `__new__` scope. Doesn't this, in combination with the register lead to problems when having more than one `Simulator`?

Reply:
I am not sure about pickling. Depends on the point at which the method is used (after obtaining the logp graph, the function is not needed anymore). 

The registration should be fine because I am creating a new `type` which has a unique name. I also have a test for multiple Simulators with different methods and it seems to be fine.

Reply:
I can also try to add a layer of indirection when creating the simulator so that I can attach what are now the local variables to the `tag`, in which case a single logp function / dispatcher would be enough.